sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionHelper.scala

```diff
@@ -0,0 +1,717 @@
+          case Some(d) => findWiderTypeWithoutStringPromotionForTwo(d, c)
+          case None => None
+        }
+    )
+  }
+
+  /**
+   * Check whether the given types are equal ignoring nullable, containsNull and valueContainsNull.
+   */
+  def haveSameType(types: Seq[DataType]): Boolean = {
+    if (types.size <= 1) {
+      true
+    } else {
+      val head = types.head
+      types.tail.forall(e => DataTypeUtils.sameType(e, head))
+    }
+  }
+
+  protected def castIfNotSameType(expr: Expression, dt: DataType): Expression = {
+    if (!DataTypeUtils.sameType(expr.dataType, dt)) {
+      Cast(expr, dt)
+    } else {
+      expr
+    }
+  }
+
+  protected def findWiderDateTimeType(d1: DatetimeType, d2: DatetimeType): DatetimeType =
+    (d1, d2) match {
+      case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
+        TimestampType
+
+      case (_: TimestampType, _: TimestampNTZType) | (_: TimestampNTZType, _: TimestampType) =>
+        TimestampType
+
+      case (_: TimestampNTZType, _: DateType) | (_: DateType, _: TimestampNTZType) =>
+        TimestampNTZType
+    }
+
+  /**
+   * Type coercion helper that matches agaist [[In]] and [[InSubquery]] expressions in order to
+   * type coerce LHS and RHS to expected types.
+   */
+  object InTypeCoercion {
+    def apply(expression: Expression): Expression = expression match {
+      // Handle type casting required between value expression and subquery output
+      // in IN subquery.
+      case i @ InSubquery(lhs, l: ListQuery) if !i.resolved && lhs.length == l.plan.output.length =>
+        // LHS is the value expressions of IN subquery.
+        // RHS is the subquery output.
+        val rhs = l.plan.output
+
+        val commonTypes = lhs.zip(rhs).flatMap {
+          case (l, r) =>
+            findWiderTypeForTwo(l.dataType, r.dataType)
+        }
+
+        // The number of columns/expressions must match between LHS and RHS of an
+        // IN subquery expression.
+        if (commonTypes.length == lhs.length) {
+          val castedRhs = rhs.zip(commonTypes).map {
+            case (e, dt) if e.dataType != dt => Alias(Cast(e, dt), e.name)()
+            case (e, _) => e
+          }
+          val newLhs = lhs.zip(commonTypes).map {
+            case (e, dt) if e.dataType != dt => Cast(e, dt)
+            case (e, _) => e
+          }
+
+          InSubquery(newLhs, l.withNewPlan(Project(castedRhs, l.plan)))
+        } else {
+          i
+        }
+
+      case i @ In(a, b) if b.exists(_.dataType != a.dataType) =>
+        findWiderCommonType(i.children.map(_.dataType)) match {
+          case Some(finalDataType) => i.withNewChildren(i.children.map(Cast(_, finalDataType)))
+          case None => i
+        }
+
+      case other => other
+    }
+  }
+
+  /**
+   * Type coercion helper that matches against function expression in order to type coerce function
+   * argument types to expected types.
+   */
+  object FunctionArgumentTypeCoercion {
```

Comment:
why are they in the helper instead of individual files?

Reply:
This is because these objects depend on methods that are defined in `Helper` but are only implemented in concrete `TypeCoercion` objects. This makes it so that the inheritance chain needs to be `Helper` -> `FunctionArgumentTypeCoercion` -> `TypeCoercion/AnsiTypeCoercion`. So I think all these helper objects need to be bunched into a single class so I just left them at the root as it seemed to be the cleanest solution. Don't really love it, so if you have any suggestions we should change it.